framework/include/ics/InitialConditionAbstract.h

```diff
@@ -0,0 +1,121 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef INITIALCONDITIONABSTRACT_H
+#define INITIALCONDITIONABSTRACT_H
+
+#include "MooseObject.h"
+#include "Coupleable.h"
+#include "FunctionInterface.h"
+#include "UserObjectInterface.h"
+#include "Restartable.h"
+#include "BlockRestrictable.h"
+#include "DependencyResolverInterface.h"
+#include "BoundaryRestrictable.h"
+#include "MooseTypes.h"
+
+// forward declarations
+class InitialConditionAbstract;
+class SystemBase;
+class MooseVariableFEBase;
+namespace libMesh
+{
+class Point;
+}
+
+template <>
+InputParameters validParams<InitialConditionAbstract>();
+
+/**
+ * InitialConditionAbstracts are objects that set the initial value of variables.
+ */
+class InitialConditionAbstract : public MooseObject,
```

Comment:
To be consistent with the rest of MOOSE, this should be named `InitialConditionBase` - we do not use `Abstract`. And `InitialConditionBase` could be renamed to `InitialConditionType` or `InitialConditionTempl` (like we had it in `SystemTempl` before it was refactored).